packages/flutter/lib/src/material/tabs.dart

```diff
@@ -1699,22 +1699,16 @@ class _TabBarState extends State<TabBar> {
     }
 
     final List<Widget> wrappedTabs = List<Widget>.generate(widget.tabs.length, (int index) {
-      const double verticalAdjustment = (_kTextAndIconTabHeight - _kTabHeight)/2.0;
-      EdgeInsetsGeometry? adjustedPadding;
-
-      if (widget.tabs[index] is PreferredSizeWidget) {
-        final PreferredSizeWidget tab = widget.tabs[index] as PreferredSizeWidget;
-        if (widget.tabHasTextAndIcon && tab.preferredSize.height == _kTabHeight) {
-          if (widget.labelPadding != null || tabBarTheme.labelPadding != null) {
-            adjustedPadding = (widget.labelPadding ?? tabBarTheme.labelPadding!).add(const EdgeInsets.symmetric(vertical: verticalAdjustment));
-          }
-          else {
-            adjustedPadding = const EdgeInsets.symmetric(vertical: verticalAdjustment, horizontal: 16.0);
```

Comment:
This refactor was quite the ordeal!

### Step 1

```dart
const EdgeInsets kTabLabelPadding = EdgeInsets.symmetric(horizontal: 16.0);
const double verticalAdjustment = (_kTextAndIconTabHeight - _kTabHeight) / 2.0;

EdgeInsetsGeometry? adjustedPadding;

final Widget tab = widget.tabs[index];
if (tab is PreferredSizeWidget) {
  if (widget.tabHasTextAndIcon && tab.preferredSize.height == _kTabHeight) {
    if (widget.labelPadding != null || tabBarTheme.labelPadding != null) {
      adjustedPadding = (widget.labelPadding ?? tabBarTheme.labelPadding!).add(
        const EdgeInsets.symmetric(vertical: verticalAdjustment),
      );
    }
    else {
      adjustedPadding = kTabLabelPadding.add(
        const EdgeInsets.symmetric(vertical: verticalAdjustment),
      );
    }
  }
}

_labelPaddings[index] = adjustedPadding ?? widget.labelPadding ?? tabBarTheme.labelPadding ?? kTabLabelPadding;
```

### Step 2

```dart
const EdgeInsets kTabLabelPadding = EdgeInsets.symmetric(horizontal: 16.0);
const double verticalAdjustment = (_kTextAndIconTabHeight - _kTabHeight) / 2.0;

EdgeInsetsGeometry adjustedPadding = widget.labelPadding ?? tabBarTheme.labelPadding ?? kTabLabelPadding;

final Widget tab = widget.tabs[index];
if (tab is PreferredSizeWidget) {
  if (widget.tabHasTextAndIcon && tab.preferredSize.height == _kTabHeight) {
    adjustedPadding = adjustedPadding.add(
      const EdgeInsets.symmetric(vertical: verticalAdjustment),
    );
  }
}

_labelPaddings[index] = adjustedPadding;
```

### Step 3

```dart
const EdgeInsets kTabLabelPadding = EdgeInsets.symmetric(horizontal: 16.0);
EdgeInsetsGeometry padding = widget.labelPadding ?? tabBarTheme.labelPadding ?? kTabLabelPadding;
const double verticalAdjustment = (_kTextAndIconTabHeight - _kTabHeight) / 2.0;

if (
  widget.tabs[index] case PreferredSizeWidget(preferredSize: Size(height: _kTabHeight))
  when widget.tabHasTextAndIcon
) {
  padding = padding.add(const EdgeInsets.symmetric(vertical: verticalAdjustment));
}

_labelPaddings[index] = padding;
```

Reply:
It's probably because I'm not used to this syntax but without reading Step2 (very good job at simplifying the logic here :+1:), I have a very hard time understanding Step3.

These refactorings are very impressive, but I think we should be very cautious to rely on advanced syntax. We can already see cases where it slows down maintenance because one have to be very aware of the new syntax to understand the logic (had one case last week where I had to read the git history to see the prerefactoring code to understand the logic).
It's also probably because I'm too old :-), but it is also because I had seen the drawbacks of smart code previously (especially in Smalltalk and Java codebases).

Reply:
Thanks @bleroux, I really appreciate the feedback!

If people are fine with a non-pattern change in this PR, I'd be fine with factoring out the if-case here :)